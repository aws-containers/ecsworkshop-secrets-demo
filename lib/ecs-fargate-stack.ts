import { App, Stack, StackProps, CfnOutput } from '@aws-cdk/core';
import { Vpc } from "@aws-cdk/aws-ec2";
import { Cluster, ContainerImage, Secret as ECSSecret } from "@aws-cdk/aws-ecs";
import { ApplicationLoadBalancedFargateService } from '@aws-cdk/aws-ecs-patterns';
import { Secret } from '@aws-cdk/aws-secretsmanager';

export interface ECSStackProps extends StackProps {
  vpc: Vpc
  dbSecretArn: string
}

export class ECSStack extends Stack {

  constructor(scope: App, id: string, props: ECSStackProps) {
    super(scope, id, props);

    const containerPort = this.node.tryGetContext("containerPort");
    const containerImage = this.node.tryGetContext("containerImage");
    const creds = Secret.fromSecretCompleteArn(this, 'postgresCreds', props.dbSecretArn);

    const cluster = new Cluster(this, 'Cluster', {
      vpc: props.vpc,
      clusterName: 'fargateClusterDemo'
    });

    const fargateService = new ApplicationLoadBalancedFargateService(this, "fargateService", {
      cluster,
      taskImageOptions: {
        image: ContainerImage.fromRegistry(containerImage),
        containerPort: containerPort,
        enableLogging: true,
        secrets: {
          POSTGRES_DATA: ECSSecret.fromSecretsManager(creds)
        }
      },
      desiredCount: 1,
      publicLoadBalancer: true,
      serviceName: 'fargateServiceDemo'
    });

    new CfnOutput(this, 'LoadBalancerDNS', { value: fargateService.loadBalancer.loadBalancerDnsName });
  }

}