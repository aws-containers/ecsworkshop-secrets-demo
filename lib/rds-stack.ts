import { App, StackProps, Stack, Duration, RemovalPolicy, CfnOutput } from "@aws-cdk/core";
import {
    DatabaseSecret, Credentials, ServerlessCluster, DatabaseClusterEngine, ParameterGroup, AuroraCapacityUnit
} from '@aws-cdk/aws-rds';
import { Vpc, Port, SubnetType } from '@aws-cdk/aws-ec2';
import { Secret, SecretRotation, SecretRotationApplication } from '@aws-cdk/aws-secretsmanager';

export interface RDSStackProps extends StackProps {
    vpc: Vpc
}

export class RDSStack extends Stack {

    readonly dbSecret: DatabaseSecret;
    readonly postgresRDSserverless: ServerlessCluster;

    constructor(scope: App, id: string, props: RDSStackProps) {
        super(scope, id, props);

        const dbUser = this.node.tryGetContext("dbUser");
        const dbName = this.node.tryGetContext("dbName");
        const dbPort = this.node.tryGetContext("dbPort") || 5432;

        this.dbSecret = new Secret(this, 'dbCredentialsSecret', {
            secretName: "ecsworkshop/test/todo-app/aurora-pg",
            generateSecretString: {
                secretStringTemplate: JSON.stringify({
                    username: dbUser,
                }),
                excludePunctuation: true,
                includeSpace: false,
                generateStringKey: 'password'
            }
        });

        this.postgresRDSserverless = new ServerlessCluster(this, 'postgresRdsServerless', {
            engine: DatabaseClusterEngine.AURORA_POSTGRESQL,
            parameterGroup: ParameterGroup.fromParameterGroupName(this, 'ParameterGroup', 'default.aurora-postgresql10'),
            vpc: props.vpc,
            enableDataApi: true,
            vpcSubnets: { subnetType: SubnetType.PRIVATE },
            credentials: Credentials.fromSecret(this.dbSecret, dbUser),
            scaling: {
                autoPause: Duration.minutes(10), // default is to pause after 5 minutes of idle time
                minCapacity: AuroraCapacityUnit.ACU_8, // default is 2 Aurora capacity units (ACUs)
                maxCapacity: AuroraCapacityUnit.ACU_32, // default is 16 Aurora capacity units (ACUs)
            },
            defaultDatabaseName: dbName,
            deletionProtection: false,
            removalPolicy: RemovalPolicy.DESTROY
        });

        this.postgresRDSserverless.connections.allowFromAnyIpv4(Port.tcp(dbPort));

        new SecretRotation(
            this,
            `ecsworkshop/test/todo-app/aurora-pg`,
            {
                secret: this.dbSecret,
                application: SecretRotationApplication.POSTGRES_ROTATION_SINGLE_USER,
                vpc: props.vpc,
                vpcSubnets: { subnetType: SubnetType.PRIVATE },
                target: this.postgresRDSserverless,
                automaticallyAfter: Duration.days(30),
            }
        );

        new CfnOutput(this, 'SecretName', { value: this.dbSecret.secretName });
    }
}