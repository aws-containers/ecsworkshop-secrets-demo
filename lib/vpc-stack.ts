import { App, Stack, StackProps, Construct } from '@aws-cdk/core';
import { Vpc, SubnetType } from '@aws-cdk/aws-ec2'

export interface VpcProps extends StackProps {
    maxAzs: number;
}

export class VPCStack extends Stack {
    readonly vpc: Vpc;

    constructor(scope: Construct, id: string, props: VpcProps) {
        super(scope, id, props);

        if (props.maxAzs !== undefined && props.maxAzs <= 1) {
            throw new Error('maxAzs must be at least 2.');
        }

        this.vpc = new Vpc(this, 'ecsWorkshopVPC', {
            cidr: "10.0.0.0/16",
            subnetConfiguration: [
                {
                    cidrMask: 24,
                    name: 'public',
                    subnetType: SubnetType.PUBLIC,
                },
                {
                    cidrMask: 24,
                    name: 'private',
                    subnetType: SubnetType.PRIVATE,
                },
            ],
        });
    }
}