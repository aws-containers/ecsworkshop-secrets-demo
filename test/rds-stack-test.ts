import { SynthUtils, expect as expectCDK, haveResource, countResources } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';
import { RDSStack } from '../lib/rds-stack';
import { VPCStack } from '../lib/vpc-stack';
import '@aws-cdk/assert/jest';

const iVpc = {
    maxAzs: 2
}

test('Fine-Grained Assertions', () => {
    const app = new App();
    // WHEN
    const vpcStack = new VPCStack(app, 'VPCStack', iVpc);
    const rdsStack = new RDSStack(app, 'RDSStack', {
        vpc: vpcStack.vpc
    })
    // THEN
    expectCDK(vpcStack).to(haveResource('AWS::EC2::VPC', {
        EnableDnsHostnames: true,
        EnableDnsSupport: true,
        InstanceTenancy: "default",
    }));
    expectCDK(rdsStack).to(haveResource('AWS::RDS::DBCluster', {
        EnableHttpEndpoint: true,
        EngineMode: "serverless",
    }));
});


test('Snapshot Tests', () => {
    const app = new App();
    const vpcStack = new VPCStack(app, 'VPCStack', iVpc);
    const rdsStack = new RDSStack(app, 'RDSStack', {
        vpc: vpcStack.vpc
    })
    expect(SynthUtils.toCloudFormation(rdsStack)).toMatchSnapshot();
});

