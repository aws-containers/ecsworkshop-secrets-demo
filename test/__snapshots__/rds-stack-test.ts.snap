// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests 1`] = `
Object {
  "Resources": Object {
    "dbCredentialsRotation79020087": Object {
      "Properties": Object {
        "Location": Object {
          "ApplicationId": "arn:aws:serverlessrepo:us-east-1:297356227824:applications/SecretsManagerRDSPostgreSQLRotationSingleUser",
          "SemanticVersion": "1.1.60",
        },
        "Parameters": Object {
          "endpoint": Object {
            "Fn::Join": Array [
              "",
              Array [
                "https://secretsmanager.",
                Object {
                  "Ref": "AWS::Region",
                },
                ".",
                Object {
                  "Ref": "AWS::URLSuffix",
                },
              ],
            ],
          },
          "functionName": "RDSStackdbCredentialsRotationC27B26BD",
          "vpcSecurityGroupIds": Object {
            "Fn::GetAtt": Array [
              "dbCredentialsRotationSecurityGroup369F1392",
              "GroupId",
            ],
          },
          "vpcSubnetIds": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::ImportValue": "VPCStack:ExportsOutputRefWorkshopVPCprivateSubnet1Subnet3A829D6F085AFE1B",
                },
                ",",
                Object {
                  "Fn::ImportValue": "VPCStack:ExportsOutputRefWorkshopVPCprivateSubnet2Subnet786F2147249A5054",
                },
              ],
            ],
          },
        },
      },
      "Type": "AWS::Serverless::Application",
    },
    "dbCredentialsRotationSecurityGroup369F1392": Object {
      "Properties": Object {
        "GroupDescription": "RDSStack/dbCredentialsRotation/SecurityGroup",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "VPCStack:ExportsOutputRefWorkshopVPCEB5EA86E11F669E4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "dbCredentialsSecretAttachment74E2BBA8": Object {
      "Properties": Object {
        "SecretId": Object {
          "Ref": "dbCredentialsSecretCA7419D4",
        },
        "TargetId": Object {
          "Ref": "postgresRdsServerlessA07A995D",
        },
        "TargetType": "AWS::RDS::DBCluster",
      },
      "Type": "AWS::SecretsManager::SecretTargetAttachment",
    },
    "dbCredentialsSecretCA7419D4": Object {
      "Properties": Object {
        "GenerateSecretString": Object {
          "ExcludePunctuation": true,
          "GenerateStringKey": "password",
          "IncludeSpace": false,
          "SecretStringTemplate": "{}",
        },
        "Name": "serverless-credentials",
      },
      "Type": "AWS::SecretsManager::Secret",
    },
    "dbCredentialsSecretPolicy9F3D95E3": Object {
      "Properties": Object {
        "ResourcePolicy": Object {
          "Statement": Array [
            Object {
              "Action": "secretsmanager:DeleteSecret",
              "Effect": "Deny",
              "Principal": Object {
                "AWS": Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":iam::",
                      Object {
                        "Ref": "AWS::AccountId",
                      },
                      ":root",
                    ],
                  ],
                },
              },
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "SecretId": Object {
          "Ref": "dbCredentialsSecretCA7419D4",
        },
      },
      "Type": "AWS::SecretsManager::ResourcePolicy",
    },
    "dbCredentialsSecretRotationScheduleB177EFD0": Object {
      "Properties": Object {
        "RotationLambdaARN": Object {
          "Fn::GetAtt": Array [
            "dbCredentialsRotation79020087",
            "Outputs.RotationLambdaARN",
          ],
        },
        "RotationRules": Object {
          "AutomaticallyAfterDays": 30,
        },
        "SecretId": Object {
          "Ref": "dbCredentialsSecretCA7419D4",
        },
      },
      "Type": "AWS::SecretsManager::RotationSchedule",
    },
    "postgresRdsServerlessA07A995D": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "DBClusterParameterGroupName": "default.aurora-postgresql10",
        "DBSubnetGroupName": Object {
          "Ref": "postgresRdsServerlessSubnets9121D013",
        },
        "DeletionProtection": false,
        "EnableHttpEndpoint": true,
        "Engine": "aurora-postgresql",
        "EngineMode": "serverless",
        "MasterUserPassword": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "dbCredentialsSecretCA7419D4",
              },
              ":SecretString:password::}}",
            ],
          ],
        },
        "MasterUsername": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{{resolve:secretsmanager:",
              Object {
                "Ref": "dbCredentialsSecretCA7419D4",
              },
              ":SecretString:username::}}",
            ],
          ],
        },
        "ScalingConfiguration": Object {
          "AutoPause": true,
          "MaxCapacity": 32,
          "MinCapacity": 8,
          "SecondsUntilAutoPause": 600,
        },
        "StorageEncrypted": true,
        "VpcSecurityGroupIds": Array [
          Object {
            "Fn::GetAtt": Array [
              "postgresRdsServerlessSecurityGroupFB8125BF",
              "GroupId",
            ],
          },
        ],
      },
      "Type": "AWS::RDS::DBCluster",
      "UpdateReplacePolicy": "Delete",
    },
    "postgresRdsServerlessSecurityGroupFB8125BF": Object {
      "Properties": Object {
        "GroupDescription": "RDS security group",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "SecurityGroupIngress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "from 0.0.0.0/0:5432",
            "FromPort": 5432,
            "IpProtocol": "tcp",
            "ToPort": 5432,
          },
        ],
        "VpcId": Object {
          "Fn::ImportValue": "VPCStack:ExportsOutputRefWorkshopVPCEB5EA86E11F669E4",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "postgresRdsServerlessSecurityGroupfromRDSStackdbCredentialsRotationSecurityGroup40A8E460IndirectPortBAFAE787": Object {
      "Properties": Object {
        "Description": "from RDSStackdbCredentialsRotationSecurityGroup40A8E460:{IndirectPort}",
        "FromPort": Object {
          "Fn::GetAtt": Array [
            "postgresRdsServerlessA07A995D",
            "Endpoint.Port",
          ],
        },
        "GroupId": Object {
          "Fn::GetAtt": Array [
            "postgresRdsServerlessSecurityGroupFB8125BF",
            "GroupId",
          ],
        },
        "IpProtocol": "tcp",
        "SourceSecurityGroupId": Object {
          "Fn::GetAtt": Array [
            "dbCredentialsRotationSecurityGroup369F1392",
            "GroupId",
          ],
        },
        "ToPort": Object {
          "Fn::GetAtt": Array [
            "postgresRdsServerlessA07A995D",
            "Endpoint.Port",
          ],
        },
      },
      "Type": "AWS::EC2::SecurityGroupIngress",
    },
    "postgresRdsServerlessSubnets9121D013": Object {
      "Properties": Object {
        "DBSubnetGroupDescription": "Subnets for postgresRdsServerless database",
        "SubnetIds": Array [
          Object {
            "Fn::ImportValue": "VPCStack:ExportsOutputRefWorkshopVPCprivateSubnet1Subnet3A829D6F085AFE1B",
          },
          Object {
            "Fn::ImportValue": "VPCStack:ExportsOutputRefWorkshopVPCprivateSubnet2Subnet786F2147249A5054",
          },
        ],
      },
      "Type": "AWS::RDS::DBSubnetGroup",
    },
  },
  "Transform": "AWS::Serverless-2016-10-31",
}
`;
